### PR TITLE
`#[doc(hidden)]` can sometimes make glob re-exports not be public API.

### PR DESCRIPTION
**Pull request by obi1kenobi**
If a glob re-export is `#[doc(hidden)]` and not deprecated, items accessible *directly* through it are not public API because removing that glob re-export shouldn't be breaking.

However, there's a subtlety: it's possible to publicly re-export the names made visible by a `#[doc(hidden)]` glob re-export, making those items public API. This results in some nasty edge cases, so we're adding lots of test cases to cover them thoroughly.

### CODE REVIEW

**Comment by obi1kenobi** [comment]
@codex review please

File: src/visibility_tracker.rs [hunk 1]
```diff
@@ -613,14 +750,16 @@ fn recursively_compute_visited_names_for_glob<'a>(
...
                 crate_,
                 module_id,
                 glob_id,
                 traversal_state,
-                visited,
-                names,
-                duplicated_names,
+                glob_resolution,
+                glob_modifiers,
+                false,
```
[the remark this answers is not in the document]
Comment (obi1kenobi) [reply src/visibility_tracker.rs:762]: This is mistaken. The inner `#[doc(hidden)]` doesn't matter here since the exported name depends only on the outer (non-hidden) re-export. The inner hidden re-export is an internal implementation detail. I had local Codex confirm this against rustdoc in HTML as well.
